### PR TITLE
class for display profile

### PR DESCRIPTION
**Pull request by Ennowulff**
I added a class to get the display profile.
Although it's only a function call, it annoys me that it's not object oriented.
plus: there are so many functionalities which are not obvious so I would like to add methods for it.
To use the context feature I added the method SET_CONTEXT where the fields of the structure used for the context will be added to the field catalog.

I also added the method SET_VALUE to set values of the display profile. I am not sure if this is a good way... I like it if attributes are public so that they can be easily changed from outside. But I know that i am lonely with this opinion... 😄 

Demo display profile
```
"Create logger
DATA(my_logger) = zcl_logger_factory=>create_log( desc = 'ABAP Logger Demo 03' ) ##no_text.

"create display profile
DATA(my_profile) = zcl_logger_factory=>create_display_profile( )->set_grid( abap_true ).
TRY.
    my_profile->set_value( i_fld = 'SHOW_ALL' i_val = abap_true ).
    my_profile->set_value( i_fld = 'NONSENSE' i_val = 'Does not matter' ) ##no_text.
  CATCH zcx_logger_display_profile INTO DATA(error).
    my_logger->e( error->get_text( ) ).
ENDTRY.


"put some messages
my_logger->i( 'program start' ) ##no_text.
my_logger->w( 'This is a warning' ) ##no_text.
my_logger->s( 'Everything ok at the end' ) ##no_text.
my_logger->s( 'Log saved. See transaction SLG1' ) ##no_text.

"Display messages in popup
my_logger->popup( my_profile->get( ) ).

```
Demo set context

```
"create display profile
DATA(my_profile) = zcl_logger_factory=>create_display_profile(
               i_single_log  = abap_true )->set_grid( abap_true ).
my_profile->set_context( 'ZLOGGER_DEMO_03_CONTEXT_S' ).
my_profile->set_value(
    i_fld = 'EXP_LEVEL'
    i_val = 0 ).
```

### CODE REVIEW

File: src/zcl_logger_display_profile.clas.abap [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  METHOD zif_logger_display_profile~set_context.
+    CHECK display_profile IS NOT INITIAL.
+
+    CLEAR display_profile-lev1_fcat.
+    CLEAR display_profile-lev1_sort.
+    CLEAR display_profile-lev2_fcat.
+    CLEAR display_profile-lev2_sort.
+
+    DATA(strucdescr) = CAST cl_abap_structdescr( cl_abap_structdescr=>describe_by_name( i_context_structure ) ).
+    DATA(colpos) = 100.
```
[the remark this answers is not in the document]
Comment (Ennowulff) [reply src/zcl_logger_display_profile.clas.abap:80]: Identifiers should be lower case: "DATA"

Suggested change 
`DATA(colpos) = 100.`
`data(colpos) = 100.`

@larshp I think this is not right...?!

**Comment by vidyadharg** [comment]
Hi [Ennowulff](https://github.com/Ennowulff),

  I have checked your solution and it worked to the fullest.

**Comment by mbtools** [comment]
just a suggestion: rename `zcx_logger_display_profile` to `zcx_logger` (or `zcx_logger_exception`). Then using the exception in other places won't cause headaches.

**Comment by Ennowulff** [comment]
the exception is only meant to be used in the profile class... 😄

**Comment by mbtools** [comment]
of course... every class should have its own exception class 🕺 😸

**Comment by vidyadharg** [comment]
Hi @Ennowulff,

I have adjusted the [Demos](https://github.com/vidyadharg/ABAP-Logger-Demos). Please check the output with your code.

level_fcat 1, 2, 3.. tables are used to set the top/ left side tree. should not be clear Itab lev1_fcat, lev1_sort.
context field should be filled in mess_fcat. the caller will fill level fcats as per requirement.

I suggest the following changes in method zcl_logger_display_profile->set_context.


```
  
METHOD zif_logger_display_profile~set_context.
    CHECK display_profile IS NOT INITIAL.

    DATA strucdescr TYPE REF TO cl_abap_structdescr.
    DATA colpos TYPE i VALUE 100.
    DATA sortpos TYPE i VALUE 1.

    strucdescr ?= cl_abap_structdescr=>describe_by_name( i_context_structure ).

    DATA component TYPE abap_compdescr.
    DATA mess_fcat LIKE LINE OF display_profile-mess_fcat.

    LOOP AT strucdescr->components INTO component.

      CLEAR mess_fcat.

      mess_fcat-ref_table = i_context_structure.
      mess_fcat-ref_field = component-name.
      mess_fcat-col_pos   = colpos.
      APPEND mess_fcat TO display_profile-mess_fcat.

      colpos = colpos + 1.

    ENDLOOP.

    colpos = colpos + 1.
    mess_fcat-ref_table = 'BAL_S_SHOW'.
    mess_fcat-ref_field = 'T_MSGTY'.
    mess_fcat-col_pos   = colpos.
    APPEND mess_fcat TO display_profile-mess_fcat.

  ENDMETHOD.


```
Report ZDEMO_LOGGER04_STANDARD output with your code,
image

output with suggested changes,
image

**Comment by Ennowulff** [comment]
Hey @vidyadharg thanks for checking!
Regarding the context fields in the tree level: 
I agree that in your example using different logs it does not make sense to group by context fields.
The idea behind it is, to group fields for better overview.
Maybe we should add an option where to add the context fields? or use two methods:
* set_context_mess
* set_context_tree (with option to display context fields on level1 or to use log as level1 and add context fields to level2

**Comment by vidyadharg** [comment]
Hi @Ennowulff,

good to create two methods.

refer program which creates tree,  [zdemo_logger04_self.prog.abap](https://github.com/vidyadharg/ABAP-Logger-Demos/blob/main/src/zdemo_logger04_self.prog.abap).

Vidyadhar.

**Comment by Ennowulff** [comment]
I split set_content into two methods:
set_content_message
set_content_tree
```
"create display profile
DATA(my_profile) = zcl_logger_factory=>create_display_profile(
               i_single_log  = abap_true )->set_grid( abap_true ).
my_profile->set_context_message( 'ZLOGGER_DEMO_03_CONTEXT_S' ).
CASE 'X'.
  WHEN p_notree.
  WHEN p_cxtree.
    my_profile->set_context_tree( 'ZLOGGER_DEMO_03_CONTEXT_S' ).
  WHEN p_cxlev2.
    my_profile->set_context_tree(
      i_context_structure = 'ZLOGGER_DEMO_03_CONTEXT_S'
      i_under_log         = abap_true ).
ENDCASE.
```

**Comment by Ennowulff** [comment]
don't understand @abaplint [message](https://github.com/ABAP-Logger/ABAP-Logger/pull/119/files#annotation_6298144398) 😢 
image

**Comment by mbtools** [comment]
The error is misleading but can happen if abaplint can't determine the complete syntax tree. Two things:
- Pretty-print with "keyword upper case"
- Downport the `DATA(...)` and `FIELD-SYMBOL(<...>)` parts to use local variables (since it should run on 7.02)

File: src/zcl_logger_display_profile.clas.abap [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+    CHECK display_profile IS NOT INITIAL.
+
+    DATA colpos    TYPE i VALUE 100.
+    DATA sortpos   TYPE i VALUE 1.
+    DATA mess_fcat LIKE LINE OF display_profile-mess_fcat.
+    DATA component TYPE cl_abap_structdescr=>component.
+
+    LOOP AT get_structure_components( i_context_structure ) INTO component.
+
```
Comment (mbtools) [inline src/zcl_logger_display_profile.clas.abap:79]: needs a local variable to hold result of `get_structure_components`

File: src/zcl_logger_factory.clas.abap [hunk 3]
```diff
@@ -185,4 +195,14 @@ CLASS zcl_logger_factory IMPLEMENTATION.
     CREATE OBJECT r_collection TYPE zcl_logger_collection.
   ENDMETHOD.
 
+  METHOD create_display_profile.
+    CREATE OBJECT r_display_profile TYPE zcl_logger_display_profile.
+    r_display_profile->set(
```
Comment (mbtools) [inline src/zcl_logger_factory.clas.abap:200]: will go away once `zcl_logger_display_profile` has no errors

**Comment by AlexandreHT** [comment]
Hi and thanks for the PR. Having a helper class in order to facilitate the setting definition is a good idea. Here are one request and some observations open to discussion:

## Request
- Exception should be catchable with one super class for the library: could you create an zcx_logger exception and inherit zcx_logger_display_profile from it? (though I agree with @mbtools that it one exception could be enough)

## Observations
- Since the purpose of this class is to build a profil, i'd be in favor of making display_profil public and READ-ONLY and ditching the get method.
- The set method would be clearer with only one parameter relying on an enumeration as constants (cleanABAP favors enumeration class but it seems a bit too much on this occasion). 
- I think we'd be better off if set_value were private: it would force us to create methods with meaningful name instead of using a generic set_value. 
- A bit more controversial: I'm not sure using a static factory for creating a zcl_display_profile is worth it. Generally value object don't need it (and the fact that you put all the logic the set method corroborates it. Why not moving it to a constructor? (OK, because it helps with the interface, but all the same: I usually don't use interfaces for helper class... ^^)) What do you think?

**Comment by AlexandreHT** [comment]
Hi, 
Let's drop my observations. 
@Ennowulff : Could you create a header exception `zcx_logger` and inherit zcx_logger_display_profile from it + correct the linter errors so we can merge the PR?

**Comment by Ennowulff** [comment]
Sorry, @AlexandreHT 
I wasn't mentioned about your comment or simply forgot... 😣
I'll do the changes ASAP

**Review by AlexandreHT — APPROVED** [review]

**Comment by AlexandreHT** [comment]
Thanks

**Pull request merged**